### PR TITLE
Switch to emscripten upstream backend.

### PR DESCRIPTION
**Pull request by ekpyron**

### CODE REVIEW

File: cmake/EthCompilerSettings.cmake [hunk 1]
```diff
@@ -16,9 +16,11 @@
...
 include(EthCheckCXXCompilerFlag)
 
-eth_add_cxx_compiler_flag_if_supported(-fstack-protector-strong have_stack_protector_strong_support)
-if(NOT have_stack_protector_strong_support)
-	eth_add_cxx_compiler_flag_if_supported(-fstack-protector)
+if(NOT EMSCRIPTEN)
+	eth_add_cxx_compiler_flag_if_supported(-fstack-protector-strong have_stack_protector_strong_support)
+	if(NOT have_stack_protector_strong_support)
+		eth_add_cxx_compiler_flag_if_supported(-fstack-protector)
+	endif()
```
Comment (ekpyron) [inline cmake/EthCompilerSettings.cmake:23]: These cause linker errors for upstream emscripten builds (``wasm-ld: error: CMakeFiles/soljson.dir/libsolc.cpp.o: undefined symbol: __stack_chk_guard
 ``). If we're bothered by this we can consider https://github.com/emscripten-core/emscripten/blob/1536301d7934663025c973e1b2e302365b9d8679/src/settings.js#L80

File: cmake/EthCompilerSettings.cmake [hunk 2]
```diff
@@ -109,15 +111,13 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 			# Re-enable exception catching (optimisations above -O1 disable it)
 			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s DISABLE_EXCEPTION_CATCHING=0")
 			# Remove any code related to exit (such as atexit)
-			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s NO_EXIT_RUNTIME=1")
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s EXIT_RUNTIME=0")
```
Comment (ekpyron) [inline cmake/EthCompilerSettings.cmake:114]: Seems the proper names are the ones without ``NO_`` (see https://github.com/emscripten-core/emscripten/blob/master/src/settings.js)

File: cmake/EthCompilerSettings.cmake [hunk 3]
```diff
@@ -109,15 +111,13 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
...
 			# Remove any code related to filesystem access
-			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s NO_FILESYSTEM=1")
-			# Remove variables even if it needs to be duplicated (can improve speed at the cost of size)
-			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s AGGRESSIVE_VARIABLE_ELIMINATION=1")
```
Comment (ekpyron) [inline cmake/EthCompilerSettings.cmake:116]: This one is marked "fastcomp-only", so no longer applies to us.

File: scripts/travis-emscripten/build_emscripten.sh [hunk 4]
```diff
@@ -54,6 +54,9 @@ cmake \
   -DTESTS=0 \
   ..
 make -j 4 soljson
+# Patch soljson.js for backwards compatibility.
+# TODO: remove this with 0.7.
+sed -i -e 's/addFunction(func,sig){/addFunction(func,sig){sig=sig||"viiiii";/' libsolc/soljson.js
```
Comment (axic) [inline scripts/travis-emscripten/build_emscripten.sh:60]: Perhaps explain this is the signature of the callback function and link to the solc-js PR.

File: .circleci/docker/Dockerfile.emscripten [hunk 5]
```diff
@@ -47,6 +45,7 @@ RUN set -ex; \
 		-DZ3_BUILD_TEST_EXECUTABLES=OFF \
 		-DZ3_BUILD_EXECUTABLE=OFF \
 		-DZ3_SINGLE_THREADED=ON \
+		-DCMAKE_CXX_FLAGS="-s DISABLE_EXCEPTION_CATCHING=0" \
```
Comment (ekpyron) [inline .circleci/docker/Dockerfile.emscripten:48]: We should think about which other flags we might want to take over from our settings to here.

**Comment by ekpyron** [comment]
In case enabling exceptions in Z3 works out now, we could actually still consider shaping off those 6MB of exception handling code of Z3 by just catching ``...`` as well as ``z3::exception`` around z3 calls (apparently this is what the flag does: exceptions *are* still thrown but without a proper exception object)... but it may be too dangerous - if there's any ``catch`` in Z3 that needs to be working, this would break... EDIT: they *do* catch exceptions by type, so this is out of the question.

File: scripts/travis-emscripten/build_emscripten.sh (hunk 4 above)
Comment (ekpyron) [reply scripts/travis-emscripten/build_emscripten.sh:60]: Added a comment - but there is no solc-js PR yet for this, is there :-)? I could create one, though.

Comment (ekpyron) [reply scripts/travis-emscripten/build_emscripten.sh:60]: https://github.com/ethereum/solc-js/pull/470

**Comment by ekpyron** [comment]
Just rebased it.

**Review by chriseth — APPROVED** [review]

**Pull request merged**